compiler/rustc_mir_transform/src/shim.rs

```diff
@@ -644,8 +644,11 @@ fn build_call_shim<'tcx>(
     let sig = sig.map_bound(|sig| tcx.erase_late_bound_regions(sig));
 
     assert_eq!(sig_substs.is_some(), !instance.has_polymorphic_mir_body());
-    let mut sig =
-        if let Some(sig_substs) = sig_substs { sig.subst(tcx, &sig_substs) } else { sig.0 };
+    let mut sig = if let Some(sig_substs) = sig_substs {
+        sig.subst(tcx, &sig_substs)
+    } else {
+        sig.skip_binder()
```

Comment:
subst_identity i feel

Reply:
I would expect this to be `no_bound_vars()` :thinking: though that may break with polymorphization rn?

Reply:
This type of pattern was always sketchy to me. Adding the EarlyBinder type makes it clear why.

I'm not sure how polymorphization works; seems like this could be a byproduct of it.

Reply:
Yeah, this panics with `no_bound_vars().unwrap()` unfortunately.

From reading the [polymorphization section](https://rustc-dev-guide.rust-lang.org/backend/monomorph.html#polymorphization) in the rustc_dev_guide, it looks like polymorphization uses `ty::Param` to represent unused generics. So after polymorphization `sig` may appear to `needs_subst()` because of the presence of `ty::Param`s. I'm sure it's more complicated than that in practice, but is that sort of the right way to think about this?

Reply:
> Yeah, this panics with `no_bound_vars().unwrap()` unfortunately.
> 
> From reading the [polymorphization section](https://rustc-dev-guide.rust-lang.org/backend/monomorph.html#polymorphization) in the rustc_dev_guide, it looks like polymorphization uses `ty::Param` to represent unused generics. So after polymorphization `sig` may appear to `needs_subst()` because of the presence of `ty::Param`s. I'm sure it's more complicated than that in practice, but is that sort of the right way to think about this?

yeah, it is exactly that, I want to move polymorphization to change instances to `ty::Binder` and use bound vars instead of params (as params are conceptionally just a different representation of bound vars or placeholders, depending on context :grin:)

I think using `ty::Param` for polymorphization will end up causing issues like the above in the future.

Reply:
Another option would be to introduce `ty::Erased`, right? Similar to `ReErased` for lifetimes.

Reply:
Both `for<T, U> my_function<T, U>` and `for<T> my_function<T, T>` should be equal to be able to codegen a single instance, right? Polymorphization already makes sure that it won't erase types that actually matter for codegen.

Reply:
> Both `for<T, U> my_function<T, U>` and `for<T> my_function<T, T>` should be equal to be able to codegen a single instance, right?

yes, the issue is that if you polymorphize to bound vars, these two are different, but if you polymorphize to erased, then both end up as `my_function<erased, erased>` which should be equal to itself

Reply:
You can't observe the difference at runtime, right? If you did, polymorphization should reject the candidates. Rustc passes `unnamed_addr` to LLVM, so just comparing the function pointers can return true already if after optimizations the function bodies are identical.

Reply:
`for<T, U> my_function<T, U>` and `for<T> my_function<T, T>` probably won't happen in practice but you could imagine
```rust
fn check_type_ids<T, U, V>() -> u8 {
    if TypeId::of::<T>() == TypeId::of::<U>() {
        0
    } else if TypeId::of::<T>() == TypeId::of::<U>() {
        1
    } else {
        2
    }
}
```
we end up with `for<Eq, DontCare> check_type_ids<Eq, DontCare, Eq>` and `for<DontCare, Eq> check_type_ids<DontCare, Eq, Eq>`
```rust
fn check_type_ids<Eq, DontCare, Eq>() -> u8 {
    0
}

fn check_type_ids<DontCare, Eq, Eq>() -> u8 {
    if TypeId::of::<T>() == TypeId::of::<U>() {
        0
    } else {
        1
    }
}
```
this isn't a great example which is why I said "While we may not necessarily end up with these cases in practice it still feels a bit dangerous". I am worried about having yet another global invariants which is difficult to sensibly check